Implement support for the Clear-Site-Data HTTP header

cdumez opened the pull request:
#### c65efd2b205ccaef913dff861a6e354dbdb9a212
<pre>
Implement support for the Clear-Site-Data HTTP header
<a href="https://bugs.webkit.org/show_bug.cgi?id=203215">https://bugs.webkit.org/show_bug.cgi?id=203215</a>

Reviewed by Darin Adler.

Implement support for the Clear-Site-Data HTTP header:
- <a href="https://w3c.github.io/webappsec-clear-site-data">https://w3c.github.io/webappsec-clear-site-data</a>

All header values are implemented expect for &quot;executionContexts&quot;.

Firefox and Chrome already ship this.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/executionContexts.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/executionContexts.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation-insecure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation-insecure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/resource-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/resource.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/storage.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/storage.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/controlled-endpoint.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/echo-clear-site-data.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/iframe_executionContexts.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/page_using_service_worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/page_with_resource.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/send_report.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/service_worker.js: Added.
(e.url.pathname.match):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/test_utils.sub.js: Added.
(TestUtils.STORAGE.string_appeared_here):
(string_appeared_here):
(TestUtils.DATATYPES.string_appeared_here):
(TestUtils.COMBINATIONS):
(TestUtils.getClearSiteDataUrl):
(TestUtils.getPageWithResourceUrl):
(TestUtils):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/w3c-import.log: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/platform/network/HTTPHeaderNames.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigins):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
(WebKit::NetworkResourceLoader::didReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::computeNetworkProcessAccessTypeForDataFetch):
(WebKit::computeWebProcessAccessTypeForDataFetch):
(WebKit::computeNetworkProcessAccessTypeForDataRemoval):
(WebKit::WebsiteDataStore::computeWebProcessAccessTypeForDataRemoval):
(WebKit::computeWebProcessAccessTypeForDataRemoval): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigins):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/254745@main">https://commits.webkit.org/254745@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/820491ae8c495e6f3c577726a757a412cbc5afd3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90046 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99379 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156884 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33085 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93675 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95694 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26299 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76856 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26203 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81158 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69204 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15036 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15977 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3334 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38938 "Passed tests") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35061 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -127,6 +127,8 @@ class NetworkConnectionToWebProcess
     void transferKeptAliveLoad(NetworkResourceLoader&);
     void setOnLineState(bool);
 
+    void deleteWebsiteDataForOrigins(OptionSet<WebsiteDataType>, const Vector<WebCore::SecurityOriginData>&, CompletionHandler<void()>&&);
```
Can we pass a `Span` instead of a `const Vector&`?

cdumez replied:
Not easily. Our current IPC encoder/decode for Span wouldn't work for any non POD type. It encodes/decodes everything as a `const uint8_t*`.

darinadler commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -392,6 +393,8 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
     void setEmulatedConditions(PAL::SessionID, std::optional<int64_t>&& bytesPerSecondLimit);
 #endif
 
+    void deleteWebsiteDataForOrigins(PAL::SessionID, OptionSet<WebsiteDataType>, const Vector<WebCore::SecurityOriginData>& origins, const Vector<String>& cookieHostNames, const Vector<String>& HSTSCacheHostnames, const Vector<RegistrableDomain>&, CompletionHandler<void()>&&);
```
Can we pass a `Span` instead of a `const Vector&`?

cdumez committed:
Implement support for the Clear-Site-Data HTTP header
https://bugs.webkit.org/show_bug.cgi?id=203215

Reviewed by Darin Adler.

Implement support for the Clear-Site-Data HTTP header:
- https://w3c.github.io/webappsec-clear-site-data

All header values are implemented expect for "executionContexts".

Firefox and Chrome already ship this.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/executionContexts.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/executionContexts.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation-insecure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation-insecure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/navigation.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/resource-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/resource.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/storage.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/storage.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/controlled-endpoint.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/echo-clear-site-data.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/iframe_executionContexts.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/page_using_service_worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/page_with_resource.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/send_report.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/service_worker.js: Added.
(e.url.pathname.match):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/test_utils.sub.js: Added.
(TestUtils.STORAGE.string_appeared_here):
(string_appeared_here):
(TestUtils.DATATYPES.string_appeared_here):
(TestUtils.COMBINATIONS):
(TestUtils.getClearSiteDataUrl):
(TestUtils.getPageWithResourceUrl):
(TestUtils):
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/w3c-import.log: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/platform/network/HTTPHeaderNames.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigins):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
(WebKit::NetworkResourceLoader::didReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::computeNetworkProcessAccessTypeForDataFetch):
(WebKit::computeWebProcessAccessTypeForDataFetch):
(WebKit::computeNetworkProcessAccessTypeForDataRemoval):
(WebKit::WebsiteDataStore::computeWebProcessAccessTypeForDataRemoval):
(WebKit::computeWebProcessAccessTypeForDataRemoval): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigins):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/254745@main
